// noarg/dist/schema/TypeNumber.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeNumber = void 0;
const TypeCore_1 = require("./TypeCore.cjs");
const result_1 = require("./result.cjs");
class TypeNumber extends TypeCore_1.TypeCore {
    constructor() {
        super(...arguments);
        this.name = 'number';
    }
    checkType(value) {
        if (value === '')
            return new result_1.ResultErr(`Number can not be empty string`);
        const number = typeof value === 'string' ? Number(value) : value;
        if (isNaN(number))
            return new result_1.ResultErr(`\`${value}\` is not a valid number`);
        if (this.config.min && number < this.config.min) {
            return new result_1.ResultErr(`Number must be at least ${this.config.min}`);
        }
        if (this.config.max && number > this.config.max) {
            return new result_1.ResultErr(`Number must be at most ${this.config.max}`);
        }
        if (this.config.enum &&
            this.config.enum.size &&
            !this.config.enum.has(number)) {
            return new result_1.ResultErr(`\`${number}\` is not in enum`);
        }
        return new result_1.ResultOk(this.config.toInteger ? Math.floor(number) : number);
    }
    min(min) {
        this.config.min = min;
        return this;
    }
    max(max) {
        this.config.max = max;
        return this;
    }
    toInteger() {
        this.config.toInteger = true;
        return this;
    }
}
exports.TypeNumber = TypeNumber;
