// noarg/dist/schema/TypeCore.mjs
import { ResultErr } from "./result.mjs";
import validateFlagName from "../helpers/validate-flag-name.mjs";
export class TypeCore {
    constructor(config) {
        this.config = config;
        this.name = 'core';
    }
    parse(value) {
        const result = this.checkType(value);
        if (result instanceof ResultErr) {
            return { value: null, error: result.message, valid: false };
        }
        return { value: result.value, error: null, valid: true };
    }
    checkType(_) {
        return new ResultErr("This type doesn't have a checkType method");
    }
    default(value) {
        this.config.required = true;
        this.config.default = value;
        return this;
    }
    ask(askQuestion) {
        this.config.required = true;
        this.config.askQuestion = askQuestion ?? 'Enter a value:';
        return this;
    }
    aliases(...aliases) {
        aliases.forEach((alias) => {
            validateFlagName(alias);
        });
        this.config.aliases = [...new Set(aliases)].sort((a, b) => a.length - b.length);
        return this;
    }
    required() {
        this.config.required = true;
        return this;
    }
    description(description) {
        this.config.description = description;
        return this;
    }
}
